projects/igniteui-angular/src/lib/grids/grid-mrl-navigation.service.ts

```diff
@@ -0,0 +1,466 @@
+
+        // previous element is from the same layout
+        let prevElementColumn = columnLayout.children
+        .find(c => (c.colEnd === currentColStart || c.colStart + c.gridColumnSpan === currentColStart ) &&
+            c.rowStart <= currentRowStart &&
+            (currentRowStart < c.rowEnd || currentRowStart < c.rowStart + c.gridRowSpan));
+
+        let columnIndex = columnLayout.children.toArray().indexOf(prevElementColumn);
+        let prevElement = element.children[columnIndex];
+
+        if (!prevElement) {
+            // try extracting first element from the previous layout
+            const prevLayout = this.grid.columns.find(c => c.columnLayout && !c.hidden && c.visibleIndex === columnLayout.visibleIndex - 1);
+            if (!prevLayout) {
+                // reached the end
+                return null;
+            }
+            const layoutSize = prevLayout.getInitialChildColumnSizes(prevLayout.children).length;
+            // first element is from the next layout
+            prevElementColumn = prevLayout.children
+            .find(c => (c.colEnd === layoutSize + 1 || c.colStart + c.gridColumnSpan === layoutSize + 1) &&
+                c.rowStart <= currentRowStart &&
+                (currentRowStart < c.rowEnd || currentRowStart < c.rowStart + c.gridRowSpan));
+
+            columnIndex = prevLayout.children.toArray().indexOf(prevElementColumn);
+
+            if (!this.isColumnLeftFullyVisible(prevElementColumn.visibleIndex)) {
+                this.grid.nativeElement.focus({ preventScroll: true });
+                this.grid.parentVirtDir.onChunkLoad
+                .pipe(first())
+                .subscribe(() => {
+                    prevElement = element.previousElementSibling.children[columnIndex];
+                    this._focusCell(prevElement);
+                });
+                const hScroll = this.horizontalScroll(cell.rowIndex);
+                const scrIndex = hScroll.igxForOf.indexOf(prevElementColumn.parent);
+                hScroll.scrollTo(scrIndex);
+                return;
+            } else {
+                prevElement = !element.previousElementSibling && this.grid.pinnedColumns.length ?
+                    element.parentNode.previousElementSibling.children[columnIndex] :
+                    element.previousElementSibling.children[columnIndex];
+            }
+        }
+        this._focusCell(prevElement);
+    }
+
+    public onKeydownEnd(rowIndex, isSummary = false, cellRowStart?) {
+        const layouts = this.grid.columns.filter(c => c.columnLayout && !c.hidden).length;
+        const lastLayout = this.grid.columns.filter(c => c.columnLayout && !c.hidden)[layouts - 1];
+        const lastLayoutChildren = lastLayout.children;
+        const layoutSize =  lastLayout.getInitialChildColumnSizes(lastLayoutChildren).length;
+        const currentRowStart =  cellRowStart || this.grid.multiRowLayoutRowSize;
+        const nextElementColumn = lastLayout.children.find(c =>
```

Comment:
these finds are very likely repeated in other places. e.g. wouldn't End look like shift + tab from the start? At least regarding columns. It just looks like each function has calculations on its own and the calculations are extremely similar from one another yet repeated everytime.

Reply:
Tab/Shift+Tab works on visible indexes only, so there the check is simpler - just gets the largest visible index and passes it to the base navigateUp function.